File: frontend/src/types/agent.ts
import type { MasteryStatus } from "./problem";

export type AgentRole = "user" | "assistant";
export type AgentToolStatus =
  | "succeeded"
  | "failed"
  | "pending_confirmation"
  | "confirmed";
export type LearningMemoryType =
  | "misconception"
  | "strength"
  | "review_focus";

export interface AgentMessage {
  id: number;
  role: AgentRole;
  content: string;
  sequence: number;
  created_at: string;
}

export interface AgentToolCall {
  id: number;
  trigger_message_id: number;
  tool_name: string;
  result_summary: string;
  status: AgentToolStatus;
  duration_ms: number;
  proposed_mastery_status: MasteryStatus | null;
  proposed_personal_notes: string | null;
  created_at: string;
  confirmed_at: string | null;
}

export interface LearningMemory {
  id: number;
  memory_type: LearningMemoryType;
  content: string;
  created_at: string;
  updated_at: string;
}

export interface AgentConversation {
  problem_id: number;
  session_id: number | null;
  messages: AgentMessage[];
  tool_calls: AgentToolCall[];
  memories: LearningMemory[];
  mastery_status: MasteryStatus;
  personal_notes: string;
}

File: backend/app/schemas/agent.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

from app.schemas.problem import MasteryStatus

AgentMessageContent = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=4000),
]
AgentRole = Literal["user", "assistant"]
AgentToolStatus = Literal[
    "succeeded",
    "failed",
    "pending_confirmation",
    "confirmed",
]
LearningMemoryType = Literal[
    "misconception",
    "strength",
    "review_focus",
]


class AgentMessageRequest(BaseModel):
    content: AgentMessageContent


class AgentMessageView(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    role: AgentRole
    content: str
    sequence: int
    created_at: datetime


class AgentToolCallView(BaseModel):
    id: int
    trigger_message_id: int
    tool_name: str
    result_summary: str
    status: AgentToolStatus
    duration_ms: int = Field(ge=0)
    proposed_mastery_status: MasteryStatus | None = None
    proposed_personal_notes: str | None = None
    created_at: datetime
    confirmed_at: datetime | None


class LearningMemoryView(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    memory_type: LearningMemoryType
    content: str
    created_at: datetime
    updated_at: datetime


class AgentConversation(BaseModel):
    problem_id: int
    session_id: int | None
    messages: list[AgentMessageView]
    tool_calls: list[AgentToolCallView]
    memories: list[LearningMemoryView]
    mastery_status: MasteryStatus
    personal_notes: str
